
initializeGenomeObject <- function(file, fasta=TRUE, expression.file=NULL, append=FALSE)
{
  genome <- new(Genome, 1, "ROC", TRUE)
  if (fasta == TRUE)
  {
    genome$readFasta(file, append)
  }
  else
  {
    genome$readRFPFile(file)
  }
  if(!is.null(expression.file))
  {
    genome$readObservedPhiValues(expression.file, FALSE)
  }
  return(genome)
}

length.Rcpp_Genome <- function(genome)
{
  return(genome$getGenomeSize())
}

summary.Rcpp_Genome <- function(genome)
{
  # TODO output stuff like:
  # - no. of genes
  # - avg. gene length
  # - avg. A,C,G,T content
  # - avg. AA composition
  # - ...
}